lib/model/losses.py

```diff
@@ -509,46 +481,52 @@ def scharr_edges(image, magnitude):
 
     # Define vertical and horizontal Scharr filters.
     static_image_shape = image.get_shape()
-    image_shape = tf.shape(image)
-
-    # modified 3x3 Scharr
-    # kernels = [[[-17.0, -61.0, -17.0], [0.0, 0.0, 0.0], [17.0, 61.0, 17.0]],
-    #         [[-17.0, 0.0, 17.0], [-61.0, 0.0, 61.0], [-17.0, 0.0, 17.0]]]
-
-    # 5x5 Scharr
-    kernels = [[[-1.0, -2.0, -3.0, -2.0, -1.0],
-                [-1.0, -2.0, -6.0, -2.0, -1.0],
-                [0.0, 0.0, 0.0, 0.0, 0.0],
-                [1.0, 2.0, 6.0, 2.0, 1.0],
-                [1.0, 2.0, 3.0, 2.0, 1.0]],
-               [[-1.0, -1.0, 0.0, 1.0, 1.0],
-                [-2.0, -2.0, 0.0, 2.0, 2.0],
-                [-3.0, -6.0, 0.0, 6.0, 3.0],
-                [-2.0, -2.0, 0.0, 2.0, 2.0],
-                [-1.0, -1.0, 0.0, 1.0, 1.0]]]
-    num_kernels = len(kernels)
-    kernels = np.transpose(np.asarray(kernels), (1, 2, 0))
-    kernels = np.expand_dims(kernels, -2) / np.sum(np.abs(kernels))
-    kernels_tf = tf.constant(kernels, dtype=image.dtype)
-    kernels_tf = tf.tile(kernels_tf, [1, 1, image_shape[-1], 1], name='scharr_filters')
+    image_shape = K.shape(image)
+
+    # 5x5 modified Scharr kernel ( reshape to (5,5,1,2) )
+    matrix = [[[[0.00070, 0.00070]],
+               [[0.00520, 0.00370]],
+               [[0.03700, 0.00000]],
+               [[0.00520, -0.0037]],
+               [[0.00070, -0.0007]]],
+              [[[0.00370, 0.00520]],
+               [[0.11870, 0.11870]],
+               [[0.25890, 0.00000]],
+               [[0.11870, -0.1187]],
+               [[0.00370, -0.0052]]],
+              [[[0.00000, 0.03700]],
+               [[0.00000, 0.25890]],
+               [[0.00000, 0.00000]],
+               [[0.00000, -0.2589]],
+               [[0.00000, -0.0370]]],
+              [[[-0.0037, 0.00520]],
+               [[-0.1187, 0.11870]],
+               [[-0.2589, 0.00000]],
+               [[-0.1187, -0.1187]],
+               [[-0.0037, -0.0052]]],
+              [[[-0.0007, 0.00070]],
+               [[-0.0052, 0.00370]],
+               [[-0.0370, 0.00000]],
+               [[-0.0052, -0.0037]],
+               [[-0.0007, -0.0007]]]]
+    num_kernels = [2]
+    kernels = K.constant(matrix, dtype='float32')
+    kernels = K.tile(kernels, [1, 1, image_shape[-1], 1])
+
 
     # Use depth-wise convolution to calculate edge maps per channel.
+    # Output tensor has shape [batch_size, h, w, d * num_kernels].
     pad_sizes = [[0, 0], [2, 2], [2, 2], [0, 0]]
     padded = tf.pad(image, pad_sizes, mode='REFLECT')
```

Comment:
Could this be changed to use plaidml pad or tf. pad depending on backend please ?
See top of layers.py for this.